### ISSUE TITLE
System.Data.SqlClient.SqlException using recurringjob with .NET Core Entity Framework

### ISSUE THREAD

**Issue by meoz** [HangfireIO/Hangfire#1811]
Hi All,

We have implemented Hangfire with .NET Core web API on Kestrel. The Hangfire database is integrated in our code first model. with Entity Framework 3.1.

Hangfire logs the following exception:

Any indication what could be the cause of this  ?

> {
>   "Depth": 0,
>   "ClassName": "System.Data.SqlClient.SqlException",
>   "Message": "A transport-level error has occurred when receiving results from the server. (provider: Session Provider, error: 19 - Physical connection is not usable)",
>   "Source": "Core .Net SqlClient Data Provider",
>   "StackTraceString": "   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)\r\n   at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)\r\n   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)\r\n   at System.Data.SqlClient.TdsParserStateObject.ThrowExceptionAndWarning(Boolean callerHasConnectionLock, Boolean asyncClose)\r\n   at System.Data.SqlClient.TdsParserStateObject.ReadSniError(TdsParserStateObject stateObj, UInt32 error)\r\n   at System.Data.SqlClient.TdsParserStateObject.ReadSniSyncOverAsync()\r\n   at System.Data.SqlClient.TdsParserStateObject.TryReadNetworkPacket()\r\n   at System.Data.SqlClient.TdsParserStateObject.TryPrepareBuffer()\r\n   at System.Data.SqlClient.TdsParserStateObject.TryReadByte(Byte& value)\r\n   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)\r\n   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()\r\n   at System.Data.SqlClient.SqlDataReader.get_MetaData()\r\n   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString)\r\n   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, SqlDataReader ds)\r\n   at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior)\r\n   at Dapper.SqlMapper.ExecuteReaderWithFlagsFallback(IDbCommand cmd, Boolean wasClosed, CommandBehavior behavior)\r\n   at Dapper.SqlMapper.QueryMultipleImpl(IDbConnection cnn, CommandDefinition& command)\r\n   at Dapper.SqlMapper.QueryMultiple(IDbConnection cnn, String sql, Object param, IDbTransaction transaction, Nullable`1 commandTimeout, Nullable`1 commandType)\r\n   at Hangfire.SqlServer.SqlServerJobQueue.<>c__DisplayClass10_0.<DequeueUsingSlidingInvisibilityTimeout>b__1(DbConnection connection)\r\n   at Hangfire.SqlServer.SqlServerStorage.UseConnection[T](DbConnection dedicatedConnection, Func`2 func)\r\n   at Hangfire.SqlServer.SqlServerJobQueue.DequeueUsingSlidingInvisibilityTimeout(String[] queues, CancellationToken cancellationToken)\r\n   at Hangfire.SqlServer.SqlServerJobQueue.Dequeue(String[] queues, CancellationToken cancellationToken)\r\n   at Hangfire.SqlServer.SqlServerConnection.FetchNextJob(String[] queues, CancellationToken cancellationToken)\r\n   at Hangfire.Server.Worker.Execute(BackgroundProcessContext context)\r\n   at Hangfire.Server.BackgroundProcessDispatcherBuilder.ExecuteProcess(Guid executionId, Object state)\r\n   at Hangfire.Processing.BackgroundExecution.Run(Action`2 callback, Object state)",
>   "RemoteStackTraceString": null,
>   "RemoteStackIndex": 0,
>   "HResult": -2146232060,
>   "HelpURL": null
> }

**Further technical details**
Hanfire version: 1.7.18
EF Core version: 3.1
Entity Framework 3.1

**Comment by odinserj** [comment]
"Physical connection is not usable" error message is likely means there's a bug in the System.Data.SqlClient package itself, and there were some cases like this in the past. Please try to upgrade or install the latest version explicitly to your application's project, the latest version of `System.Data.SqlClient` package is [`4.8.2`](https://www.nuget.org/packages/System.Data.SqlClient/4.8.2).

It's not bumped by Hangfire.SqlClient automatically, because some applications are still depend on old versions, and version bump is a breaking change that can cause more troubles than it solves.

**Comment by meoz** [comment]
Hangfire.SqlServer 1.7.18 does use System.Data.SqlClient package is 4.8.2.
![image](https://user-images.githubusercontent.com/332372/107535335-ee400d00-6bc0-11eb-8568-a68fac92d14c.png)

**Comment by odinserj** [comment]
No, it would be ObjectDisposedException in this case. Have you tried to google the error message? I've just found this post – https://stackoverflow.com/a/32364278
